addons/zylann.hterrain/native/src/quad_tree_lod.cpp

```diff
@@ -0,0 +1,219 @@
+
+    int lod_factor = get_lod_factor(lod);
+    int chunk_size = _base_size * lod_factor;
+    Vector3 world_center = (real_t)chunk_size * (Vector3((real_t)quad->origin_x, 0, (real_t)quad->origin_y) + Vector3(0.5, 0, 0.5));
+
+    if (_vertical_bounds_func.is_valid()) {
+        Variant ret = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
+        Vector2 vbounds;
+        if (ret.get_type() == Variant::VECTOR2)
+            vbounds = (Vector2)ret;
+        else
+            Godot::print("Error: QuadtreeLod::_update: _vertical_bounds_func did not return a vector2");
+        world_center.y = (vbounds.x + vbounds.y) / 2.0;
+    }
+
+    int split_distance = _base_size * lod_factor * _split_scale;
+
+    if (! quad->has_children()) {
+        if (lod > 0 && world_center.distance_to(view_pos) < split_distance) {
+            // Split
+            quad->add_children();
+
+            for (int i = 0; i < 4; i++) {
+                Quad* child = new Quad();
+                child->origin_x = quad->origin_x * 2 + (i & 1);
+                child->origin_y = quad->origin_y * 2 + ((i & 2) >> 1);
+                child->data = _make_chunk(lod - 1, child->origin_x, child->origin_y);
+                quad->set_child(i, child);
+                // If the quad needs to split more, we'll ask more recycling...
+            }
+
+            if (quad->data.get_type() == Variant::OBJECT) {
+                _recycle_chunk(quad->data, quad->origin_x, quad->origin_y, lod);
+                quad->data = Variant();
+            }
+        }
+    } else {
+        bool no_split_child = true;
+
+        for (int i = 0; i < 4; i++) {
+            _update(quad->get_child(i), lod - 1, view_pos);
+            if (quad->get_child(i)->has_children())
+                no_split_child = false;
+        }
+
+        if (no_split_child && world_center.distance_to(view_pos) > split_distance) {
+            // Join
+            if (quad->has_children()) {
+                for (int i = 0; i < 4; i++) {
+                    Quad* child = quad->get_child(i);
+                    _recycle_chunk(child->data, child->origin_x, child->origin_y, lod - 1);
+                    quad->data = Variant();
+                }
+                quad->clear_children();
+
+                //assert(quad->data == null);
+                //if ( quad->data.is_valid() )    // Really necessary after it was just unrefed in quad?
+                	//Godot::print("QuadTreeLod error: quad->data is still valid");
+
+                quad->data = _make_chunk(lod, quad->origin_x, quad->origin_y);
+            }
+        }
+    }
+} // _update
+
+void QuadTreeLod::_join_all_recursively(Quad *quad, int lod) {
+    if (quad->has_children()) {
+        for (int i = 0; i < 4; i++) {
+            _join_all_recursively(quad->get_child(i), lod - 1);
+        }
+        quad->clear_children();
+
+    } else if (quad->data.get_type() == Variant::OBJECT) {
```

Comment:
The original check was != null

Reply:
`!= null`, aka `is valid`. Since data is a Variant, it's always valid.
I changed these to `quad->data.get_type() != Variant::NIL` but no difference.